### PR TITLE
Identity tp rewrite

### PR DESCRIPTION
**Pull request by eloaverona**
Add identity transaction processor in rust
Adapt identity-tp in rust so it can compile to wasm and be used as a Sabre contract

### CODE REVIEW

File: families/identity/sawtooth_identity_rust/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+[package]
+name = "sawtooth_identity_rust"
+version = "0.1.0"
+authors = ["Intel Corporation"]
```
Comment (agunde406) [inline families/identity/sawtooth_identity_rust/Cargo.toml]: Change to Bitwise io

File: families/identity/sawtooth_identity_rust/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# Copyright 2018 Cargill Incorporated
```
Comment (agunde406) [inline families/identity/sawtooth_identity_rust/Dockerfile]: Bitiwse io

File: families/identity/sawtooth_identity_rust/build.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2018 Intel Corporation
```
Comment (agunde406) [inline families/identity/sawtooth_identity_rust/build.rs]: Bitwise io

File: families/identity/sawtooth_identity_rust/src/handler.rs [hunk 4]
```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright 2018 Intel Corporation
```
Comment (agunde406) [inline families/identity/sawtooth_identity_rust/src/handler.rs]: Bitwise io

File: families/identity/sawtooth_identity_rust/src/handler.rs [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+const ROLE_PREFIX: &str = "01";
+const ALLOWED_SIGNER_SETTING: &str = "sawtooth.identity.allowed_keys";
+
+// Constants to be used when constructing config namespace addresses
+const SETTING_NAMESPACE: &str = "000000";
+const _SETTING_MAX_KEY_PARTS: usize = 4;
+const _SETTING_ADDRESS_PART_SIZE: usize = 16;
+
+
+///Computes the address for the given setting key.
```
Comment (agunde406) [inline families/identity/sawtooth_identity_rust/src/handler.rs]: ///Computes -> /// Computes

File: families/identity/sawtooth_identity_rust/src/handler.rs [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+    let mut addr_parts: Vec<_> = key_parts.iter().map(|x| setting_short_hash(x)).collect();
+    let len_addr_parts = addr_parts.len();
+    addr_parts.extend(
+        vec![get_setting_address_padding();_SETTING_MAX_KEY_PARTS - &len_addr_parts].iter().cloned());
+    SETTING_NAMESPACE.to_string() + &addr_parts.join("")
+}
+
+fn setting_short_hash(byte_str: &str) -> String {
+    // Computes the SHA 256 hash and truncates to be the length
+    // of an address part (see _config_key_to_address for information on
```
Comment (agunde406) [inline families/identity/sawtooth_identity_rust/src/handler.rs]: Unfinished comment

**Comment by agunde406** [comment]
I would also suggest you create an IdentityState to handle settign and getting state and IdentityPayload objects to handle validating the provided payload. This helps make the actually transaction processor logic easier to read. You can look at intkey rust for an example.

File: bin/identity-tp [hunk 7]
```diff
@@ -30,7 +30,7 @@ sys.path.insert(0, os.path.join(
     os.path.dirname(os.path.dirname(os.path.realpath(__file__))),
     'families', 'identity'))
 
-from sawtooth_identity.processor.main import main
+from sawtooth_identity_py.processor.main import main
```
Comment (peterschwarz) [inline bin/identity-tp]: We should avoid naming packages based on the implementation language.  This should remain as `sawtooth_identity`, but probably needs to be organized at a higher level. 

Likewise, we should talk about whether or not we want to keep the python implementation at all.

**Comment by rbuysse** [comment]
bin/build_debs and bin/build_rust will be removed by #1729. The tp should be added to docker-compose-installed.yaml and docker/compose/copy-debs.yaml instead.

**Comment by rbuysse** [comment]
bin/build_debs and bin/build_rust will be removed by #1729. The tp should be added to docker-compose-installed.yaml and docker/compose/copy-debs.yaml instead.

File: families/identity/sawtooth_identity/src/handler.rs [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+/// joined into a single address, with the config namespace
+/// (_SETTING_NAMESPACE) added at the beginning.
+///
+/// Args:
+///     key (str): the setting key
+/// Returns:
+///     str: the computed address
+///
+
+fn setting_key_to_address(key: &str) -> String {
```
Comment (peterschwarz) [inline families/identity/sawtooth_identity/src/handler.rs]: This might be a clearer way to implement this: https://github.com/hyperledger/sawtooth-core/blob/master/validator/src/state/settings_view.rs#L128-L145

File: families/identity/sawtooth_identity/src/handler.rs [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+        Ok(())
+    }
+
+
+    fn get_policy_address(&mut self, policy_name: &str) -> String {
+            format!("{}{}{}", IDENTITY_NAMESPACE.to_string(), IdentityState::POLICY_PREFIX.to_string(),
+                    self.to_hash(policy_name)[..62].to_string())
+        }
+
+    fn get_role_address(&mut self, role_name: &str) -> String {
```
Comment (peterschwarz) [inline families/identity/sawtooth_identity/src/handler.rs]: Likewise, possibly a clearer way to represent this: https://github.com/hyperledger/sawtooth-core/blob/master/validator/src/state/identity_view.rs#L182-L207

**Comment by eloaverona** [comment]
@peterschwarz @agunde406 updated

File: sdk/rust/src/processor/handler.rs [hunk 10]
```diff
@@ -389,7 +389,7 @@ impl TransactionContext {
         let x: &[u8] = &serialized;
 
         let mut future = self.sender.send(
-            Message_MessageType::TP_RECEIPT_ADD_DATA_REQUEST,
```
Comment (agunde406) [inline sdk/rust/src/processor/handler.rs]: This fix should be broken out in its own commit

File: bin/identity-tp [hunk 11]
```diff
@@ -13,24 +13,28 @@
...
-
-sys.path.insert(0, os.path.join(
-    os.path.dirname(os.path.dirname(os.path.realpath(__file__))),
-    'sdk', 'python'))
-sys.path.insert(0, os.path.join(
-    os.path.dirname(os.path.dirname(os.path.realpath(__file__))),
-    'families', 'identity', 'python'))
-
-from sawtooth_identity.processor.main import main
+# Copyright 2018 Bitwise IO, Inc.
```
Comment (agunde406) [inline bin/identity-tp]: This line should go below `Copyright 2017 Intel Corporation` and the rest of this license should be removed

File: families/identity/Dockerfile [hunk 12]
```diff
@@ -11,8 +11,26 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+# ------------------------------------------------------------------------------
 
-# docker build -f families/identity/rust/sawtooth_identity/Dockerfile -t sawtooth-identity-tp-rust-local .
+# Copyright 2018 Bitwise IO Inc.
```
Comment (agunde406) [inline families/identity/Dockerfile]: Same fix as above for the license

File: families/identity/Dockerfile-installed [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+# Copyright 2018 Bitwise IO Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (agunde406) [inline families/identity/Dockerfile-installed]: Same license fix as above

File: sdk/rust/src/processor/handler.rs (hunk 10 above)
Comment (eloaverona) [reply sdk/rust/src/processor/handler.rs]: Addressed.

File: families/identity/rust/sawtooth_identity/src/handler.rs [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+    context: &mut TransactionContext,
+) -> Result<(), ApplyError> {
+    let header = transaction.get_header();
+
+    let entries_list = get_state_data(&get_allowed_signer_address(), context)?;
+
+    match entries_list {
+        None => Err(ApplyError::InvalidTransaction(format!(
+            "The transaction signer is
+                                                      not authorized to submit transactions: {:?}",
```
Comment (agunde406) [inline families/identity/rust/sawtooth_identity/src/handler.rs]: There is weird formating here

File: families/identity/rust/sawtooth_identity/src/handler.rs [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+            for entry in setting.get_entries().iter() {
+                if entry.get_key() == ALLOWED_SIGNER_SETTING {
+                    let allowed_signer: Vec<&str> = entry.get_value().split(",").collect();
+                    if allowed_signer.contains(&header.get_signer_public_key()) {
+                        return Ok(());
+                    }
+                }
+            }
+            Err(ApplyError::InvalidTransaction(format!(
+                "The transaction signer is
```
Comment (agunde406) [inline families/identity/rust/sawtooth_identity/src/handler.rs]: this is weird formatting.

Comment (eloaverona) [reply families/identity/rust/sawtooth_identity/src/handler.rs]: updated!

**Comment by eloaverona** [comment]
See #1926
